libmscore/figuredbass.cpp

```diff
@@ -1770,6 +1773,256 @@ FiguredBass* Score::addFiguredBass()
+            int harmony;
+            switch ( modifier ) {
+                  case FiguredBassItem::Modifier::NONE: {
+                        Note *antecedent = nullptr;
+                        int antecedent_tick = -1; // if we fail to find an antecedence, this will remain -1.
+                        // Here we must apply the accidental seen earlier in the measure if there is one, even if in a different octave.
+                        // find the note with maximum tick, but <= note->tick() whose line % 7 == native_pitch->line%7, take its accidental.
+                        chord->score()->visitNotesInStaff(chord->measure(), [&](Note *n) {
+
+                              if ( n->track() != chord->track())
+                                    return;
+
+                              // we have to skip notes which are not the bottom note of the chord.
+                              // why? because these notes were auto-generated, and their note->line() returns 0 even if not really on line 0.
+                              if ( find_if(n->chord()->notes().begin(),
+                                           n->chord()->notes().end(),
+                                           [n](Note *j) {
+                                                 // if there is a note j whose pitch is < n's pitch, then n is not the bottom note, so return.
+                                                 return j->pitch() < n->pitch();
+                                           }) != n->chord()->notes().end())
+                              {
+                                    return; // *n is not the bottom note of the chord so skip it.
+                              }
+                              if ( mod7(n->line()) != mod7(harmony_line) ) // skip if it is not same line or +/- octaves
+                                    return;
+                              if ( n->chord()->tick() > chord->tick() ) // skip if it is to the right
+                                    return;
+                              if ( n->chord()->tick() < antecedent_tick)
+                                    return;
+                              antecedent = n;
+                              antecedent_tick = n->chord()->tick();
+                        });
+                        if (antecedent_tick > -1){  // if found an antecedent.
+                              harmony = note->ppitch() + mod12(antecedent->ppitch() - note->ppitch());
+                        } else {
+                              harmony = native_pitch + int(acciv);
+                        }
+                  }
+                        break;
+                  case FiguredBassItem::Modifier::DOUBLEFLAT:
+                        harmony = native_pitch + int(AccidentalVal::FLAT2);
+                        break;
+                  case FiguredBassItem::Modifier::FLAT:
+                        harmony = native_pitch + int(AccidentalVal::FLAT);
+                        break;
```

Comment:
Yes this is indeed strange.  It is the pattern already used elsewhere in the muescore source.  The AccidentalVal enum values have been
contrived so they can be converted to int and added to a pitch.   In my opinion they + should be somehow overloaded so you can just add
an integer to a AccidentalVal.  But that has not been done, and I don’t know whether in C++ you can overload + for (int + custom_type).
Anyway this is how musescore does it.  I’m just following the pattern.


> On 24 May 2015, at 05:12, Maurizio M. Gavioli <notifications@github.com> wrote:
> 
> In libmscore/figuredbass.cpp <https://github.com/musescore/MuseScore/pull/2015#discussion_r30949794>:
> 
> > +                              antecedent = n;
> > +                              antecedent_tick = n->chord()->tick();
> > +                        });
> > +                        if (antecedent_tick > -1){  // if found an antecedent.
> > +                              harmony = note->ppitch() + mod12(antecedent->ppitch() - note->ppitch());
> > +                        } else {
> > +                              harmony = native_pitch + int(acciv);
> > +                        }
> > +                  }
> > +                        break;
> > +                  case FiguredBassItem::Modifier::DOUBLEFLAT:
> > +                        harmony = native_pitch + int(AccidentalVal::FLAT2);
> > +                        break;
> > +                  case FiguredBassItem::Modifier::FLAT:
> > +                        harmony = native_pitch + int(AccidentalVal::FLAT);
> > +                        break;
> At this point, I am not sure I understand anything, but is it wise to rely on specific values for the enum AccidentalVal?
> 
> —
> Reply to this email directly or view it on GitHub <https://github.com/musescore/MuseScore/pull/2015/files#r30949794>.
>